File: languages/go/blur-par.go
// blur-par: parallel 3x3 Gaussian blur, PASSES double-buffered passes.
// Invocation: blur-par <cores> <n>
// Same algorithm and checksum as serial blur; parallelises within each pass by
// partitioning the NxN output image into `cores` contiguous row-bands.
// Workers read the full input buffer (src) — read-only, no data race — and write
// only their own rows of the output buffer (dst). After each pass the buffers are
// swapped (main goroutine) before the next pass begins. Border clamping (edge-
// replication) is identical to the serial version.
package main

import (
	"fmt"
	"os"
	"runtime"
	"strconv"
	"sync"
	"time"
)

const (
	P      = 1000000007
	PASSES = 4
)

func lcg(s int64) int64 { return (s*1103515245 + 12345) & 0x7fffffff }

func clampi(x, n int) int {
	if x < 0 {
		return 0
	}
	if x >= n {
		return n - 1
	}
	return x
}

// blurBand writes rows [rowStart, rowEnd) of dst from the full src buffer.
func blurBand(src, dst []int32, n, rowStart, rowEnd int) {
	K := [9]int32{1, 2, 1, 2, 4, 2, 1, 2, 1} // 3x3 Gaussian, sum 16
	for i := rowStart; i < rowEnd; i++ {
		for j := 0; j < n; j++ {
			var acc int32 = 0
			for di := -1; di <= 1; di++ {
				ni := clampi(i+di, n)
				for dj := -1; dj <= 1; dj++ {
					nj := clampi(j+dj, n)
					acc += K[(di+1)*3+(dj+1)] * src[ni*n+nj]
				}
			}
			dst[i*n+j] = acc / 16 // integer floor division
		}
	}
}

func run(cores, n int) int64 {
	src := make([]int32, n*n)
	dst := make([]int32, n*n)
	s := int64(42)
	for k := 0; k < n*n; k++ {
		s = lcg(s)
		src[k] = int32(s % 256)
	}

	var wg sync.WaitGroup
	t0 := time.Now()
	for pass := 0; pass < PASSES; pass++ {
		wg.Add(cores)
		for w := 0; w < cores; w++ {
			w := w // capture
			rowStart := w * n / cores
			rowEnd := (w + 1) * n / cores
			// Capture src/dst by pointer value at loop time — they are
			// resliced each pass after the swap below.
			curSrc := src
			curDst := dst
			go func() {
				defer wg.Done()
				blurBand(curSrc, curDst, n, rowStart, rowEnd)
			}()
		}
		wg.Wait()
		src, dst = dst, src // double-buffer swap (same as serial)
	}
	fmt.Fprintf(os.Stderr, "COMPUTE_NS %d\n", time.Since(t0).Nanoseconds())

	var h int64 = 0
	for k := 0; k < n*n; k++ {
		h = (h*31 + int64(src[k])) % P
	}
	return h
}

func main() {
	cores := 1
	n := 256
	if len(os.Args) > 1 {
		if v, err := strconv.Atoi(os.Args[1]); err == nil {
			cores = v
		}
	}
	if len(os.Args) > 2 {
		if v, err := strconv.Atoi(os.Args[2]); err == nil {
			n = v
		}
	}
	runtime.GOMAXPROCS(cores)
	fmt.Println(run(cores, n))
	fmt.Printf("blur(%d)\n", n)
}

File: languages/scala/blur-par.scala
// blur-par: parallel version of the blur benchmark.
// Decomposes each pass into `cores` row bands. Workers read the full input buffer
// (including neighbour rows for the 3x3 stencil — read-only, no contention) and
// write only their own output rows. After all workers complete a pass the buffers
// are swapped (same double-buffer scheme as serial) and the next pass begins.
// Border clamping: edge-replication, identical to the serial spec.
// Core-invariant: each output pixel depends only on a 3x3 region of the input;
// workers write disjoint output rows; result is bit-identical for any core count.
// Primitive: java.util.concurrent fixed-thread-pool.
// Invocation: blur-par <cores> <n>
import java.util.concurrent.{Callable, Executors}

object BlurPar {
  final val MOD: Long   = 1000000007L
  final val PASSES: Int = 4

  def lcg(s: Long): Long = (s * 1103515245L + 12345L) & 0x7fffffffL

  def clampi(x: Int, n: Int): Int = if (x < 0) 0 else if (x >= n) n - 1 else x

  def main(args: Array[String]): Unit = {
    val cores  = if (args.length >= 1) args(0).toInt else 1
    val n      = if (args.length >= 2) args(1).toInt else 256
    val warmup = sys.env.get("LL_WARMUP").flatMap(_.toIntOption).getOrElse(5)

    val k = Array(1, 2, 1, 2, 4, 2, 1, 2, 1) // 3x3 kernel, sum=16

    // LCG fill — identical seed and sequence as serial. Kept as the pristine input.
    val pristineSrc = new Array[Int](n * n)
    var s = 42L
    var idx = 0
    while (idx < n * n) { s = lcg(s); pristineSrc(idx) = (s % 256).toInt; idx += 1 }

    var src = new Array[Int](n * n)
    var dst = new Array[Int](n * n)

    val pool = Executors.newFixedThreadPool(cores)

    // Build per-worker row ranges once; reuse across passes and runs.
    val rowStart = new Array[Int](cores)
    val rowEnd   = new Array[Int](cores)
    var ww = 0
    while (ww < cores) {
      rowStart(ww) = ww * n / cores
      rowEnd(ww)   = (ww + 1) * n / cores
      ww += 1
    }

    // Run one full multi-pass blur compute starting from the current src/dst state.
    // Caller must restore src/dst before each call.
    def runCompute(): Unit = {
      var pass = 0
      while (pass < PASSES) {
        // Capture current src/dst references for this pass (mutable refs swap after join).
        val curSrc = src
        val curDst = dst

        val tasks = new java.util.ArrayList[Callable[Unit]]()
        var w = 0
        while (w < cores) {
          val wIndex = w
          val rStart = rowStart(wIndex)
          val rEnd   = rowEnd(wIndex)
          tasks.add(new Callable[Unit] {
            def call(): Unit = {
              var i = rStart
              while (i < rEnd) {
                var j = 0
                while (j < n) {
                  var acc = 0
                  var di = -1
                  while (di <= 1) {
                    val ni = clampi(i + di, n)
                    var dj = -1
                    while (dj <= 1) {
                      val nj = clampi(j + dj, n)
                      acc += k((di + 1) * 3 + (dj + 1)) * curSrc(ni * n + nj)
                      dj += 1
                    }
                    di += 1
                  }
                  curDst(i * n + j) = acc / 16
                  j += 1
                }
                i += 1
              }
            }
          })
          w += 1
        }
        // barrier: invokeAll blocks until all workers complete before the buffer swap.
        pool.invokeAll(tasks)

        // Double-buffer swap — same as serial (swap refs, not data).
        val t = src; src = dst; dst = t
        pass += 1
      }
    }

    // Reset src/dst to pristine state before each run (blur mutates src via double-buffer swap).
    def resetBuffers(): Unit = {
      System.arraycopy(pristineSrc, 0, src, 0, n * n)
      // dst is always fully overwritten before it's read; zeroing is not strictly required
      // but ensures a clean state for consistency.
      java.util.Arrays.fill(dst, 0)
    }

    // Warmup: run PASSES blur passes `warmup` times, discard results.
    var wi = 0
    while (wi < warmup) {
      resetBuffers()
      runCompute()
      wi += 1
    }

    // Timed run — restore pristine src before timing.
    resetBuffers()
    val t0 = System.nanoTime()
    runCompute()
    System.err.println("COMPUTE_NS " + (System.nanoTime() - t0))

    pool.shutdown()

    // Serial checksum over the final buffer (src after PASSES swaps).
    var h = 0L
    var p = 0
    while (p < n * n) { h = (h * 31 + src(p)) % MOD; p += 1 }

    println(h)
    println(s"blur($n)")
  }
}
